source/src/ble/cs_CrownstoneCentral.cpp

```diff
@@ -0,0 +1,663 @@
+	if (retCode != ERR_SUCCESS) {
+		return retCode;
+	}
+
+	controlPacketAccessor.setProtocolVersion(CS_CONNECTION_PROTOCOL_VERSION);
+	controlPacketAccessor.setType(commandType);
+	retCode = controlPacketAccessor.setPayload(data, size);
+	if (retCode != ERR_SUCCESS) {
+		return retCode;
+	}
+	cs_data_t controlPacket = controlPacketAccessor.getSerializedBuffer();
+
+	uint16_t encryptedSize = ConnectionEncryption::getEncryptedBufferSize(controlPacket.len, ConnectionEncryptionType::CTR);
+	if (_opMode == OperationMode::OPERATION_MODE_SETUP) {
+		ConnectionEncryption::getInstance().encrypt(controlPacket, encryptedBuffer, SETUP, ConnectionEncryptionType::CTR);
+	}
+	else {
+		ConnectionEncryption::getInstance().encrypt(controlPacket, encryptedBuffer, ADMIN, ConnectionEncryptionType::CTR);
+	}
+
+	retCode = BleCentral::getInstance().write(_controlHandle, encryptedBuffer.data, encryptedSize);
+	if (retCode != ERR_WAIT_FOR_SUCCESS) {
+		return retCode;
+	}
+
+	_currentOperation = Operation::WRITE;
+	setStep(WriteControlSteps::WRITE);
+	return ERR_WAIT_FOR_SUCCESS;
+}
+
+cs_data_t CrownstoneCentral::requestWriteBuffer() {
+	if (isBusy()) {
+		return cs_data_t();
+	}
+	cs_data_t writeBuf = CharacteristicWriteBuffer::getInstance().getBuffer();
+	ControlPacketAccessor<> controlPacketAccessor;
+	cs_ret_code_t retCode = controlPacketAccessor.assign(writeBuf.data, writeBuf.len);
+	if (retCode != ERR_SUCCESS) {
+		return cs_data_t();
+	}
+	return cs_data_t(controlPacketAccessor.getPayload().data, controlPacketAccessor.getMaxPayloadSize());
+}
+
+//////////////////// Helpers ////////////////////
+
+void CrownstoneCentral::enableNotifications() {
+	cs_ret_code_t retCode = BleCentral::getInstance().writeNotificationConfig(_resultCccdHandle, true);
+	if (retCode != ERR_WAIT_FOR_SUCCESS) {
+		finalizeOperation(Operation::CONNECT, retCode);
+		return;
+	}
+	setStep(ConnectSteps::ENABLE_NOTIFICATIONS);
+}
+
+void CrownstoneCentral::readSessionData() {
+	cs_ret_code_t retCode;
+	switch (_opMode) {
+		case OperationMode::OPERATION_MODE_NORMAL: {
+			// Read session data.
+			retCode = BleCentral::getInstance().read(_sessionDataHandle);
+			if (retCode != ERR_WAIT_FOR_SUCCESS) {
+				finalizeOperation(Operation::CONNECT, retCode);
+				return;
+			}
+			setStep(ConnectSteps::SESSION_DATA);
+			break;
+		}
+		case OperationMode::OPERATION_MODE_SETUP: {
+			// Read session key.
+			retCode = BleCentral::getInstance().read(_sessionKeyHandle);
+			if (retCode != ERR_WAIT_FOR_SUCCESS) {
+				finalizeOperation(Operation::CONNECT, retCode);
+				return;
+			}
+			setStep(ConnectSteps::SESSION_KEY);
+			break;
+		}
+		default: {
+			finalizeOperation(Operation::CONNECT, ERR_WRONG_MODE);
+			return;
+		}
+	}
+}
+
+
+//////////////////// Keeping up operation and step ////////////////////
+
+bool CrownstoneCentral::isBusy() {
+	return _currentOperation != Operation::NONE;
+}
+
+void CrownstoneCentral::setStep(ConnectSteps step) {
+	_currentStep = static_cast<uint8_t>(step);
```

Comment:
I would expect something like `if(step != _currentStep +1) { /* not allowed */ return false; }`. Am I missing something :)

Reply:
Steps can be skipped sometimes. For example, if the crownstone is not in setup mode, you don't have to read the setup key.
I guess i could add a debug log when you skip a step?